pkg/activator/certificate/cache.go

```diff
@@ -50,68 +56,138 @@ type CertCache struct {
 	if err != nil {
-		return nil, fmt.Errorf("failed to get activator certificate, secret %s/%s was not found: %w. Enabling system-internal-tls requires the secret to be present and populated with a valid certificate and CA",
+		return nil, fmt.Errorf("failed to get activator certificate, secret %s/%s was not found: %w. Enabling system-internal-tls requires the secret to be present and populated with a valid certificate",
 			system.Namespace(), netcfg.ServingRoutingCertName, err)
 	}
 
-	cr.updateCache(secret)
+	cr.updateCertificate(secret)
+	cr.updateTrustPool()
 
-	secretInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+	nsSecretInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
 		FilterFunc: controller.FilterWithNameAndNamespace(system.Namespace(), netcfg.ServingRoutingCertName),
 		Handler: cache.ResourceEventHandlerFuncs{
 			UpdateFunc: cr.handleCertificateUpdate,
 			AddFunc:    cr.handleCertificateAdd,
 		},
 	})
 
+	nsConfigmapInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.ChainFilterFuncs(
+			reconciler.LabelExistsFilterFunc(networking.TrustBundleLabelKey),
+		),
+		Handler: controller.HandleAll(func(obj interface{}) {
+			cr.updateTrustPool()
+		}),
+	})
+
 	return cr, nil
 }
 
 func (cr *CertCache) handleCertificateAdd(added interface{}) {
 	if secret, ok := added.(*corev1.Secret); ok {
-		cr.updateCache(secret)
+		cr.updateCertificate(secret)
+		cr.updateTrustPool()
 	}
 }
 
-func (cr *CertCache) updateCache(secret *corev1.Secret) {
+func (cr *CertCache) handleCertificateUpdate(_, new interface{}) {
+	cr.handleCertificateAdd(new)
+	cr.updateTrustPool()
+}
+
+func (cr *CertCache) updateCertificate(secret *corev1.Secret) {
 	cr.certificatesMux.Lock()
 	defer cr.certificatesMux.Unlock()
 
 	cert, err := tls.X509KeyPair(secret.Data[certificates.CertName], secret.Data[certificates.PrivateKeyName])
 	if err != nil {
-		cr.logger.Warnw("failed to parse secret", zap.Error(err))
+		cr.logger.Warnf("failed to parse certificate in secret %s/%s: %v", secret.Namespace, secret.Name, zap.Error(err))
 		return
 	}
 	cr.certificate = &cert
+}
 
+// CA can optionally be in `ca.crt` in the `routing-serving-certs` secret
+// and/or configured using a trust-bundle via ConfigMap that has the defined label `knative-ca-trust-bundle`.
+func (cr *CertCache) updateTrustPool() {
 	pool := x509.NewCertPool()
-	block, _ := pem.Decode(secret.Data[certificates.CaCertName])
-	ca, err := x509.ParseCertificate(block.Bytes)
-	if err != nil {
-		cr.logger.Warnw("failed to parse CA", zap.Error(err))
-		return
-	}
-	pool.AddCert(ca)
+
+	cr.addSecretCAIfPresent(pool)
+	cr.addTrustBundles(pool)
+
+	// Use the trust pool in upstream TLS context
+	cr.certificatesMux.Lock()
+	defer cr.certificatesMux.Unlock()
 
 	cr.TLSConf.RootCAs = pool
 	cr.TLSConf.ServerName = certificates.LegacyFakeDnsName
```

Comment:
Do we want to drop this?

Reply:
Not sure if we'd break somebodies existing setup if we do (even if it is alpha). Ideally we finish the feature and then look at removing it.

Reply:
Can you make a follow issue for this 'cleanup' work

Reply:
--> https://github.com/knative/serving/issues/14992